### ISSUE TITLE
Error in BIOMOD_2 - [[rast] filename is empty. Provide a valid filename]

### ISSUE THREAD

**Issue by GISilvia** [biomodhub/biomod2#456]
I'm running ensemble SDM for a plant species, Robinia pseudoacacia, with global presence data from GBIF and pseudo-absence points generated randomly. All runs of my functions work until the projection of the models with function BIOMOD_Projection. It also works, but when I try to use the resulted object with other functions, for example when I try to plot the object or to use it in other functions, it does not work and returns this error: [rast] filename is empty. Provide a valid filename

**Code used:

modformat <- BIOMOD_FormatingData(resp.var= as.data.frame(b[  ,1]) , 
                                  resp.xy = b[,c("x_centroid","y_centroid")], 
                                  resp.name = "Robinia.pseudoacacia", 
                                  expl.var=b[ ,5:10], 
                                  PA.nb.rep =3,
                                  PA.nb.absences=10000, 
                                  na.rm=T) 

modformat

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.formated.data -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

dir.name =  .

sp.name =  Robinia.pseudoacacia

	 54272 presences,  0 true absences and  100000 undefined points in dataset

	 6 explanatory variables

      bio7           bio10           bio16            bio17         bulk_density     pop_density     
 Min.   : 2.00   Min.   :-8.95   Min.   :   0.4   Min.   :   0.0   Min.   : 238.8   Min.   : 0.0000  
 1st Qu.:21.90   1st Qu.:17.25   1st Qu.: 217.0   1st Qu.:  43.4   1st Qu.:1106.8   1st Qu.: 0.6376  
 Median :26.90   Median :19.85   Median : 277.5   Median : 106.3   Median :1251.1   Median : 2.5263  
 Mean   :28.17   Mean   :20.50   Mean   : 381.7   Mean   : 124.4   Mean   :1192.6   Mean   : 2.6314  
 3rd Qu.:34.10   3rd Qu.:24.55   3rd Qu.: 429.6   3rd Qu.: 174.4   3rd Qu.:1331.8   3rd Qu.: 4.0818  
 Max.   :70.50   Max.   :38.85   Max.   :6263.1   Max.   :4258.0   Max.   :1664.8   Max.   :11.4100  

**Continue of my code:

modOption <- BIOMOD_ModelingOptions(
  GLM = list( type = 'quadratic', interaction.level = 1 ), 
  GAM = list( algo = 'GAM_mgcv' , k=1, select=T) 
) 

DataSplitTable.b <- bm_CrossValidation(bm.format = modformat,
                                       balance = 'presences',
                                       k = 5, 
                                       strategy = "strat", 
                                       nb.rep = 3)

par(mfrow=c(2,3))
plot(modformat, calib.lines = DataSplitTable.b)

![Robinia pseudoacacia_modformat_e_crossval](https://github.com/biomodhub/biomod2/assets/44375686/1c9f9102-c883-42e8-9769-a84613693702)

mod_output <- BIOMOD_Modeling(
  bm.format = modformat,
  modeling.id = paste("Robinia.pseudoacacia", "Mod_V2", sep = "_"), 
  models = c("GLM", "GAM"),
  bm.options = modOption,
  CV.do.full.models = FALSE, 
  prevalence = 0.5, 
  var.import = 3, 
  metric.eval = c('TSS', 'ROC'), 
  nb.cpu = 1, 
  seed.val = NULL,
  do.progress = TRUE,
  CV.strategy = "user.defined",
  CV.user.table = DataSplitTable.b
) 

mod_output 

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.models.out -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

Modeling folder : .

Species modeled : Robinia.pseudoacacia

Modeling id : Robinia.pseudoacacia_Mod_V2

Considered variables : bio7 bio10 bio16 bio17 bulk_density pop_density

Computed Models :  Robinia.pseudoacacia_allData_RUN1_GLM Robinia.pseudoacacia_allData_RUN1_GAM 
Robinia.pseudoacacia_allData_RUN2_GLM Robinia.pseudoacacia_allData_RUN2_GAM Robinia.pseudoacacia_allData_RUN3_GLM 
Robinia.pseudoacacia_allData_RUN3_GAM Robinia.pseudoacacia_allData_RUN4_GLM Robinia.pseudoacacia_allData_RUN4_GAM 
Robinia.pseudoacacia_allData_RUN5_GLM Robinia.pseudoacacia_allData_RUN5_GAM Robinia.pseudoacacia_allData_RUN6_GLM 
Robinia.pseudoacacia_allData_RUN6_GAM Robinia.pseudoacacia_allData_RUN7_GLM Robinia.pseudoacacia_allData_RUN7_GAM 
Robinia.pseudoacacia_allData_RUN8_GLM Robinia.pseudoacacia_allData_RUN8_GAM Robinia.pseudoacacia_allData_RUN9_GLM 
Robinia.pseudoacacia_allData_RUN9_GAM Robinia.pseudoacacia_allData_RUN10_GLM Robinia.pseudoacacia_allData_RUN10_GAM

Failed Models :  none

**Continue of my code:

raster_names <- c("bio7", "bio10", "bio16", "bio17", "bulk_density", "pop_density")
eu_pres_v2 <- rast("eu_pres_v2.tif") 
names(eu_pres_v2) <- c("bio7", "bio10", "bio16", "bio17", "bulk_density", "pop_density")

ppreseu <- BIOMOD_Projection(bm.mod=mod_output,
                             new.env = eu_pres_v2, 
                             proj.name="out_projection_eu_presente",
                             build.clamping.mask=T, 
                             output.format = ".img",
                             metric.filter = "ROC", 
                             metric.binary="ROC")

plot(ppreseu)
Error: [rast] filename is empty. Provide a valid filename

##

ppreseu

-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-= BIOMOD.projection.out -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

Projection directory : ./Robinia.pseudoacacia/out_projection_eu_presente

sp.name : Robinia.pseudoacacia

expl.var.names : bio7 bio10 bio16 bio17 bulk_density pop_density

modeling.id : Robinia.pseudoacacia_Mod_V2 ( 
./Robinia.pseudoacacia/Robinia.pseudoacacia.Robinia.pseudoacacia_Mod_V2.models.out )

models.projected : 
Robinia.pseudoacacia_allData_RUN1_GLM, Robinia.pseudoacacia_allData_RUN1_GAM, Robinia.pseudoacacia_allData_RUN2_GLM, Robinia.pseudoacacia_allData_RUN2_GAM, Robinia.pseudoacacia_allData_RUN3_GLM, Robinia.pseudoacacia_allData_RUN3_GAM, Robinia.pseudoacacia_allData_RUN4_GLM, Robinia.pseudoacacia_allData_RUN4_GAM, Robinia.pseudoacacia_allData_RUN5_GLM, Robinia.pseudoacacia_allData_RUN5_GAM, Robinia.pseudoacacia_allData_RUN6_GLM, Robinia.pseudoacacia_allData_RUN6_GAM, Robinia.pseudoacacia_allData_RUN7_GLM, Robinia.pseudoacacia_allData_RUN7_GAM, Robinia.pseudoacacia_allData_RUN8_GLM, Robinia.pseudoacacia_allData_RUN8_GAM, Robinia.pseudoacacia_allData_RUN9_GLM, Robinia.pseudoacacia_allData_RUN9_GAM, Robinia.pseudoacacia_allData_RUN10_GLM, Robinia.pseudoacacia_allData_RUN10_GAM

available binary projection : Ro, ROC, ROCfilt.img

**Environment Information**

sessionInfo()
R version 4.3.0 (2023-04-21 ucrt)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 11 x64 (build 22631)

Matrix products: default

locale:
[1] LC_COLLATE=Italian_Italy.utf8  LC_CTYPE=Italian_Italy.utf8    LC_MONETARY=Italian_Italy.utf8
[4] LC_NUMERIC=C                   LC_TIME=Italian_Italy.utf8    

time zone: Europe/Rome
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] ggpubr_0.6.0             ggplot2_3.4.4            biomod2_4.2-4            maditr_0.8.3            
 [5] terra_1.7-29             reshape2_1.4.4           CoordinateCleaner_2.0-20 taxize_0.9.100          
 [9] magrittr_2.0.3           purrr_1.0.1              dplyr_1.1.2              readr_2.1.4             
[13] rgbif_3.7.8             

loaded via a namespace (and not attached):
  [1] rstudioapi_0.14        jsonlite_1.8.4         farver_2.1.1           vctrs_0.6.2           
  [5] base64enc_0.1-3        rstatix_0.7.2          webshot_0.5.5          htmltools_0.5.5       
  [9] plotrix_3.8-4          curl_5.0.0             broom_1.0.5            raster_3.6-20         
 [13] xgboost_1.7.5.1        Formula_1.2-5          pROC_1.18.2            KernSmooth_2.23-20    
 [17] htmlwidgets_1.6.2      plyr_1.8.8             zoo_1.8-12             uuid_1.1-0            
 [21] commonmark_1.9.0       whisker_0.4.1          lifecycle_1.0.4        iterators_1.0.14      
 [25] pkgconfig_2.0.3        Matrix_1.6-4           R6_2.5.1               fastmap_1.1.1         
 [29] plotmo_3.6.2           digest_0.6.31          colorspace_2.1-0       reshape_0.8.9         
 [33] mapview_2.11.0         leafem_0.2.0           crosstalk_1.2.0        labeling_0.4.3        
 [37] conditionz_0.1.0       bold_1.3.0             randomForest_4.7-1.1   fansi_1.0.4           
 [41] gbm_2.1.8.1            httr_1.4.7             abind_1.4-5            mgcv_1.8-42           
 [45] compiler_4.3.0         proxy_0.4-27           withr_2.5.2            backports_1.4.1       
 [49] carData_3.0-5          DBI_1.1.3              rgdal_1.6-7            ggsignif_0.6.4        
 [53] MASS_7.3-58.4          leaflet_2.1.2          classInt_0.4-9         oai_0.4.0             
 [57] tools_4.3.0            units_0.8-2            ape_5.7-1              TeachingDemos_2.12    
 [61] nnet_7.3-18            glue_1.6.2             satellite_1.0.4        nlme_3.1-162          
 [65] gridtext_0.1.5         grid_4.3.0             sf_1.0-13              generics_0.1.3        
 [69] maxnet_0.1.4           gtable_0.3.4           tzdb_0.4.0             class_7.3-21          
 [73] tidyr_1.3.0            data.table_1.14.8      hms_1.1.3              sp_1.6-1              
 [77] xml2_1.3.4             car_3.1-2              utf8_1.2.3             foreach_1.5.2         
 [81] pillar_1.9.0           markdown_1.7           stringr_1.5.1          splines_4.3.0         
 [85] ggtext_0.1.2           lattice_0.21-8         survival_3.5-5         tidyselect_1.2.0      
 [89] crul_1.4.0             stats4_4.3.0           xfun_0.39              stringi_1.7.12        
 [93] lazyeval_0.2.2         codetools_0.2-19       httpcode_0.3.0         tibble_3.2.1          
 [97] cli_3.6.1              rpart_4.1.19           munsell_0.5.0          Rcpp_1.0.10           
[101] PresenceAbsence_1.1.11 rnaturalearth_0.3.2    png_0.1-8              parallel_4.3.0        
[105] mda_0.5-4              scales_1.3.0           e1071_1.7-13           earth_5.3.2           
[109] crayon_1.5.2           geosphere_1.5-18       rlang_1.1.1            rgeos_0.6-3

**Comment by MayaGueguen** [comment]
Hello Silvia,

I'm guessing it might be because you asked for several transformations of your projection :eyes:

In your case, [BIOMOD_Projection](https://biomodhub.github.io/biomod2/reference/BIOMOD_Projection.html) will compute predicted probabilities over your new environmental conditions (`eu_pres_v2`), and then from this map, will produce 2 supplementary maps using `ROC` metric as threshold to obtain filtered values and binary values.

So you get 3 output maps for each of your single models. If you print your `ppreseu` object, you will have a bit of detail about that, and more precisely, if you type `myBiomodProj@proj.out`, you will see in the `link` slot the names of all the projection files created.

Then, if you do not want to open directly the raster files, you can extract these projections through the [get_predictions](https://biomodhub.github.io/biomod2/reference/getters.out.html) function for `BIOMOD.projection.out` object : 
```
## to get probabilities
get_predictions(ppreseu) 

## to get filtered probabilities
get_predictions(ppreseu, metric.filter = "ROC")

## to get binary values
get_predictions(ppreseu, metric.binary = "ROC")
```
Do not hesitate if something is still unclear for you 

Maya

**Comment by GISilvia** [comment]
Dear Maya,
thak you for prompt reply.
The misterious thing is that the same code works with difference species. The same code also works when I just change the name from "Robinia.pseudoacacia" to "falseAcacia". It does not work after changing it in "Robinia_pseudoacacia" (in this case after running BIOMOD_FormatingData I have the worning "! Response variable name was converted into Robinia.pseudoacacia"). I first tried to run the same code on a subset of the dataset without changing that name, and it returned the same error. Then I run the subset with the name changed, and I did not have the error any more:

modformat2 <- BIOMOD_FormatingData(resp.var= as.data.frame(b2[  ,1]) , 
                                  resp.xy = b2[,c("x_centroid","y_centroid")], 
                                  resp.name = "falseAcacia", 
                                  expl.var=b2[ ,5:10], 
                                  PA.nb.rep =3, 
                                  PA.nb.absences=10000, 
                                  na.rm=T) 

DataSplitTable.b2 <- bm_CrossValidation(bm.format = modformat2,
                                       balance = 'presences',
                                       k = 5, 
                                       strategy = "strat",
                                       nb.rep = 3) 

mod_output2 <- BIOMOD_Modeling(
  bm.format = modformat2,
  modeling.id = paste("falseAcacia", "Mod_V2", sep = "_"), 
  models = c("GLM", "GAM"),
  bm.options = modOption,
  CV.do.full.models = FALSE,
  prevalence = 0.5,
  var.import = 3, 
  metric.eval = c('TSS', 'ROC'), 
  nb.cpu = 1, 
  seed.val = NULL, 
  do.progress = TRUE, 
  CV.strategy = "user.defined",
  CV.user.table = DataSplitTable.b2
) 

e <- ext(11, 13, 40, 42)
it <- crop(eu_pres_v2, e)
ppreseu2 <- BIOMOD_Projection(bm.mod=mod_output2,
                             new.env = it, 
                             proj.name="out_projection_eu_presente2",
                             build.clamping.mask=T, 
                             output.format = ".img",
                             metric.filter = "ROC", 
                             metric.binary="ROC") 
plot(ppreseu2)
plotting this way I do not have the error but nothing happen (because of a limitation of my machine I think); plotting this other way i have the correct output (I did not have it with previous name and this same plotting way):

png(filename=file.path("falseAcacia", paste("falseAcacia","project_presente.png", sep="_")), width = 1000, height = 1000, pointsize = 14, bg = "white", res = NA) 
plot(ppreseu2)   
dev.off()

it returned the correct image file:

![falseAcacia_project_presente](https://github.com/biomodhub/biomod2/assets/44375686/6096ac4f-72e7-4282-975d-0c84d157177c)

I'm guessing that ".pseudoa" is a part of some temporary file name and this somehow put the system in confusion... Coult it be?

**Comment by MayaGueguen** [comment]
Aaaah it could be because the species name (`resp.name` in [BIOMOD_FormatingData](https://biomodhub.github.io/biomod2/reference/BIOMOD_FormatingData.html)) is included within your modeling id (`modeling.id` in [BIOMOD_Modeling](https://biomodhub.github.io/biomod2/reference/BIOMOD_Modeling.html)) !

I'll check that right away !

**Comment by MayaGueguen** [comment]
Actually, I did not manage to reproduce your error :slightly_frowning_face: 
Neither with the modeling id, or searching for your pattern.

**Comment by hannahlowens** [comment]
Hi Folks,
I am having the same problem as @GISilvia with several species in the genus _Columbina_. I wonder, @MayaGueguen, if the problem is having the string "bin" in the species name. When I was trying to diagnose the problem, I looked at my `BIOMOD.projection.out` file, and it said "available binary projection : Colum". I had not produced a binary projection for the species...

**Comment by MayaGueguen** [comment]
Hello @GISilvia , Hannah,

Oups... Yes actually we found out this problem at some point, and Hélène corrected it :slightly_smiling_face: 
It was pushed this summer, you can have the fix by updating your version with the 4.2-6-1 available on github with : 
```
library(devtools)
devtools::install_github("biomodhub/biomod2", dependencies = TRUE)
```
Silvia, note that this will imply some changes in the package since you were working with 4.2-4 :eyes: 
All updates can be found on the [NEWS](https://biomodhub.github.io/biomod2/articles/news.html) page.
:warning: Starting from 4.2-5 version, modeling options changed, you can find documentation and help about that in our [vignette](https://biomodhub.github.io/biomod2/articles/vignette_modelingOptions.html), [secundary examples](https://biomodhub.github.io/biomod2/articles/examples_2_secundaryFunctions.html#retrieve-modeling-options) and in [previous issues](https://github.com/biomodhub/biomod2/issues?q=modeling+options) !

Maya

**Comment by hannahlowens** [comment]
Aha, yep, that fixed it.
